Brokerages/Bitfinex/BitfinexBrokerage.cs

```diff
@@ -0,0 +1,374 @@
+using QuantConnect.Orders;
+using QuantConnect.Packets;
+using QuantConnect.Securities;
+using QuantConnect.Util;
+using RestSharp;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net;
+
+namespace QuantConnect.Brokerages.Bitfinex
+{
+    public partial class BitfinexBrokerage : BaseWebsocketsBrokerage, IDataQueueHandler
+    {
+        private readonly BitfinexSymbolMapper _symbolMapper = new BitfinexSymbolMapper();
+
+        #region IBrokerage
+        /// <summary>
+        /// Checks if the websocket connection is connected or in the process of connecting
+        /// </summary>
+        public override bool IsConnected => WebSocket.IsOpen;
+
+        /// <summary>
+        /// Places a new order and assigns a new broker ID to the order
+        /// </summary>
+        /// <param name="order">The order to be placed</param>
+        /// <returns>True if the request for a new order has been placed, false otherwise</returns>
+        public override bool PlaceOrder(Order order)
+        {
+            return SubmitOrder(GetEndpoint("order/new"), order);
+        }
+
+        /// <summary>
+        /// Updates the order with the same id
+        /// </summary>
+        /// <param name="order">The new order information</param>
+        /// <returns>True if the request was made for the order to be updated, false otherwise</returns>
+        public override bool UpdateOrder(Order order)
+        {
+            if (order.BrokerId.Count == 0)
+            {
+                throw new ArgumentNullException("BitfinexBrokerage.UpdateOrder: There is no brokerage id to be updated for this order.");
+            }
+            if (order.BrokerId.Count > 1)
+            {
+                throw new NotSupportedException("BitfinexBrokerage.UpdateOrder: Multiple orders update not supported. Please cancel and re-create.");
+            }
+
+            return SubmitOrder(GetEndpoint("order/cancel/replace"), order);
+        }
+
+        /// <summary>
+        /// Cancels the order with the specified ID
+        /// </summary>
+        /// <param name="order">The order to cancel</param>
+        /// <returns>True if the request was submitted for cancellation, false otherwise</returns>
+        public override bool CancelOrder(Order order)
+        {
+            Log.Trace("BitfinexBrokerage.CancelOrder(): {0}", order);
+
+            if (!order.BrokerId.Any())
+            {
+                // we need the brokerage order id in order to perform a cancellation
+                Log.Trace("BitfinexBrokerage.CancelOrder(): Unable to cancel order without BrokerId.");
+                return false;
+            }
+
+            LockStream();
+            var endpoint = GetEndpoint("order/cancel/multi");
+            var payload = new JsonObject();
+            payload.Add("request", endpoint);
+            payload.Add("nonce", GetNonce().ToString());
+            payload.Add("order_ids", order.BrokerId.Select(i => long.Parse(i)));
```

Comment:
LEAN does not support multi-cancel, is there a reason for not using the single cancel endpoint ?
`/v1/order/cancel`

Reply:
Hmm. Lean allows to associate multiple brokerage orders to a single lean order (order.BrokerId is List). Am I right in thinking that we need close all associated brokerage orders if lean order closed?

Reply:
I believe the BrokerId list was implemented for Tradier brokerage who would generate two orders when crossing the zero line (e.g. I am long 1 and try to go short by selling 2). Does Bitfinex behave in the same way?
No changes are required, was just curious about the reason